### DISCUSSION TITLE
Health check URL

### DISCUSSION THREAD

**Discussion by ghost** [datopian/giftless#71] [Q&A]
Hi!

Is there some health check URL, which returns 200 on HTTP GET requests (or something similar)? And if not can these be created?

Currently, the problem is that when Giftless is deployed on Kubernetes with ingress configuration set up, it requires health check URL. Requesting the root path ("/") will return 404 and thus Kubernetes regards the pod as unhealthy.

And to provide some more context: We are deploying CKAN with ckanext-authz-service plugin and Giftless into Kubernetes. CKAN is accessed via HTTPS and as a user is redirected to Giftless for a file to be uploaded, Giftless should also use HTTPS, otherwise "Mixed content blocked" is raised.

Thank you for your help!

**Comment by shevron** [comment]
Right now there is no endpoint that will *always* return 200 regardless of configuration (e.g. the `batch` API which is the only endpoint that is currently always available may require authentication and authorization, and some specific parameters). 

If you have an nginx or similar proxy in the same pod as Giftless between the Ingress and Giftless, you can easily configure nginx to serve a health check URL. This isn't perfect as there may still be cases where nginx is working properly but for some reason Giftless will not, but it's better than nothing. 

Another good option is to write (or find an existing) a custom WSGI middleware to respond to health check probes, install it in the Docker image and enable it - see https://giftless.readthedocs.io/en/latest/configuration.html#using-wsgi-middleware for more info. Such middleware, if none already exists, should probably be a few lines of code. 

Adding such functionality as a built-in capability of Giftless is possible, but I'm not sure how I feel about that. Perhaps if you have some ideas on how to do that in a clean and secure way that will also allow users to customize it (different systems may expect different "alive" responses, etc.), you can issue a PR.

**Reply by ghost** [reply]
Thank you for your reply. I went with setting up Nginx as the reverse proxy, and it worked fine. But as you mentioned the solution is not perfect, because health checks are fake and don't measure the state of the real service.

I would expect that even very basic static/unconfigurable health check URL will be better than nothing. Generally, it is much easier to change default health check paths for Kubernetes/ECS/etc. than to configure your own in Giftless. Also as Kubernetes are so popular, just returning 200 for "/"-path will cover the most of the cases.

**Reply by shevron** [reply]
I am still inclined towards a WSGI middleware (this can be created separately from Giftless and simply added via the config file), but I can also see the value in something like a `/status` endpoint that always returns OK, potentially with some useful system info such as the version of Giftless. 

Of course a PR would be most welcome :)
